src/Workspaces/Core/Portable/Workspace/Host/TemporaryStorage/TrivialTemporaryStorageService.cs

```diff
@@ -39,31 +39,31 @@ public void Dispose()
 
             public Stream ReadStream(CancellationToken cancellationToken = default)
             {
-                if (_stream == null)
+                var stream = _stream;
+                if (stream is null)
                 {
                     throw new InvalidOperationException();
                 }
 
-                _stream.Position = 0;
-                return _stream;
+                // Return a read-only view of the underlying buffer to prevent users from overwriting or directly
+                // disposing the backing storage.
+                return new MemoryStream(stream.GetBuffer(), 0, (int)stream.Length, writable: false);
```

Comment:
📝 This is the fix for the exception (separate the readable stream from backing storage)

Reply:
Wait, does this also meant that two things trying to read the same stream at the same time would have stepped on each other, with the second call rewinding the stream position on the first one to get it? That's probably been a huge source of flakiness in tests...?

Reply:
> Wait, does this also meant that two things trying to read the same stream at the same time would have stepped on each other, with the second call rewinding the stream position on the first one to get it?

Yes

> That's probably been a huge source of flakiness in tests...?

Not sure. Most tests don't rely on this implementation.